diff --git a/nodeaccess.module b/nodeaccess.module
index 19335d2..70a6f57 100644
--- a/nodeaccess.module
+++ b/nodeaccess.module
@@ -220,9 +220,11 @@ function nodeaccess_grant_tab($node) {
  * @return mixed
  */
 function nodeaccess_grants_form($form, &$form_state, $node) {
-  $form_values = $form_state['values'];
+  if (isset($form_state['values'])) {
+    $form_values = $form_state['values'];
+  }
 
-  if (!$form_values) {
+  if (!isset($form_values)) {
     $form_values = array();
     $grants = nodeaccess_get_grants($node);
     $form_values['rid'] = isset($grants['rid']) ? $grants['rid'] : array();
@@ -939,26 +941,23 @@ function nodeaccess_get_grants($node) {
   $grants = array();
 
   // Load grants by roles.
-  $query = db_select('nodeaccess', 'na');
-  $query->join('role', 'r', 'r.rid = na.gid');
-  $query->join('nodeaccess_role_alias', 'nra', 'nra.rid = r.rid');
-  $query->fields('r', array('rid', 'name'));
+  $query = db_select('role', 'r');
+  $query->leftJoin('nodeaccess_role_alias', 'nra', 'nra.rid = r.rid');
+  $query->leftJoin('node_access', 'na', 'na.gid = r.rid AND na.realm = :realm AND na.nid = :nid', array(':realm' => 'nodeaccess_rid', ':nid' => $node->nid));
+  $query->fields('r', array('rid'));
+  $query->fields('nra', array('name'));
   $query->fields('na', array('grant_view', 'grant_update', 'grant_delete'));
-  $query->condition('na.realm', 'nodeaccess_rid', '=');
-  $query->condition('na.nid', $node->nid, '=');
   $query->orderBy('nra.weight');
   $query->orderBy('nra.name');
   $result = $query->execute();
 
   while ($grant = $result->fetch()) {
-    if ($grant->name) {
-      $grants['rid'][$grant->rid] = array(
-        'name' => $grant->name,
-        'grant_view' => (bool) $grant->grant_view,
-        'grant_update' => (bool) $grant->grant_update,
-        'grant_delete' => (bool) $grant->grant_delete,
-      );
-    }
+    $grants['rid'][$grant->rid] = array(
+      'name' => empty($grant->name) ? '' : check_plain($grant->name),
+      'grant_view' => (bool) $grant->grant_view,
+      'grant_update' => (bool) $grant->grant_update,
+      'grant_delete' => (bool) $grant->grant_delete,
+    );
   }
   // Load grants by users
   $query = db_select('node_access', 'na');
